tests/cecil-tests/ApiCapitalizationTest.cs

```diff
@@ -0,0 +1,252 @@
+			"iTunesMetadataKeyUserGenre" }
+		};
+
+		Dictionary<string, HashSet<string>> allowedFields = new Dictionary<string, HashSet<string>> () {
+		};
+
+
+		Dictionary<string, HashSet<string>> allowedMethods = new Dictionary<string, HashSet<string>> () {
+			["Dlfcn"] = new HashSet<string> () { "dlopen", "dlerror", "dlsym" }
+		};
+
+		public bool IsSkip (string type, string memberName, Dictionary<string, HashSet<string>> allowed)
+		{
+			if (allowed.TryGetValue (type, out var result)) {
+				if (allowed [type].TryGetValue (memberName, out var skipped)) {
+					return true;
+				}
+			}
+
+			return Char.IsUpper (memberName [0]);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void PropertiesCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				var typeName = type.Name;
+				var c = type.Properties
+						.Where (p => p.GetMethod?.IsPublic == true || p.SetMethod?.IsPublic == true)
+						.Where (p => !IsSkip (type.Name, p.Name, allowedProperties) && !IsPropertyObsolete (p, type) && !p.IsSpecialName)
+						.Select (p => p.Name);
+				return c;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void MethodsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				var c = from m in type.Methods
+						where m.IsPublic && !IsSkip (type.Name, m.Name, allowedMethods) && !IsException (assemblyPath, m)
+						select m.Name;
+				return c;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void EventsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				var c = from e in type.Events
+						where !(Char.IsUpper (e.Name [0]))
+						select e.Name;
+				return c;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		[TestCaseSource (typeof (Helper), nameof (Helper.NetPlatformAssemblies))]
+		[Test]
+		public void FieldsCapitalizationTest (string assemblyPath)
+		{
+			Func<TypeDefinition, IEnumerable<string>> selectLambda = (type) => {
+				var c = from f in type.Fields
+						where f.IsPublic && f.IsFamilyOrAssembly && !IsSkip (type.Name, f.Name, allowedFields)
+						select f.Name;
+				return c;
+			};
+			CapitalizationTest (assemblyPath, selectLambda);
+		}
+
+		public void CapitalizationTest (string assemblyPath, Func<TypeDefinition, IEnumerable<string>> selectLambda)
+		{
+
+			if (assemblyCache.TryGetValue (assemblyPath, out var cache)) {
+				var typeDict = new Dictionary<string, string> ();
+
+				var publicTypes = cache.MainModule.Types.Where ((t) => t.IsPublic && !IsTypeObsolete (t));
+
+				foreach (var type in publicTypes) {
```

Comment:
There can be nested types inside other types, so you'll need some sort of logic to drill into those.